src/test/java/junit/org/rapidpm/frp/model/QuadTest.java

```diff
@@ -0,0 +1,110 @@
+package junit.org.rapidpm.frp.model;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.Test;
+import org.rapidpm.frp.model.Quad;
+
+public class QuadTest {
+
+  @Test
+  public void testEquals() {
+    final Quad quad = new Quad(0, 1, 2, 3);
+
+    assertTrue(quad.equals(quad));
+    assertFalse(quad.equals(null));
+    assertFalse(quad.equals(new Quad(null, 1, 2, 3)));
+    assertFalse(quad.equals(new Quad(0, null, 2, 3)));
+    assertFalse(quad.equals(new Quad(0, 1, null, 3)));
+  }
+
+  @Test
+  public void testEqualsT1() {
+    final Quad quad = new Quad(null, 1, 2, 3);
+
+    assertTrue(quad.equals(new Quad(null, 1, 2, 3)));
+    assertFalse(quad.equals(new Quad(0, 1, 2, 3)));
+  }
+
+  @Test
+  public void testEqualsT2() {
+    final Quad quad = new Quad(0, null, 2, 3);
+
+    assertTrue(quad.equals(new Quad(0, null, 2, 3)));
+    assertFalse(quad.equals(new Quad(0, 1, 2, 3)));
+  }
+
+  @Test
+  public void testEqualsT3() {
+    final Quad quad = new Quad(0, 1, null, 3);
+
+    assertTrue(quad.equals(new Quad(0, 1, null, 3)));
+    assertFalse(quad.equals(new Quad(0, 1, 2, 3)));
+  }
+
+  @Test
+  public void testEqualsT4() {
+    final Quad quad = new Quad(0, 1, 2, null);
+
+    assertTrue(quad.equals(new Quad(0, 1, 2, null)));
+    assertFalse(quad.equals(new Quad(0, 1, 2, 3)));
+  }
+
+  @Test
+  public void testHashCode() throws Exception {
+    final Quad quad = new Quad(0, 1, 2, 3);
+
+    assertEquals(1026, quad.hashCode());
+  }
+
+  @Test
+  public void testHashCodeT1() throws Exception {
+    final Quad quad = new Quad(null, 1, 2, 3);
+
+    assertEquals(1026, quad.hashCode());
+  }
+
+  @Test
+  public void testHashCodeT2() throws Exception {
+    final Quad quad = new Quad(0, null, 2, 3);
+
+    assertEquals(65, quad.hashCode());
+  }
+
+  @Test
+  public void testHashCodeT3() throws Exception {
+    final Quad quad = new Quad(0, 1, null, 3);
+
+    assertEquals(964, quad.hashCode());
+  }
+
+  @Test
+  public void testHashCodeT4() throws Exception {
```

Comment:
Are all of these `throws Exception`s needed?

Reply:
Fixed